---
title: "Adding to the story: Annotations, maps and interactions"
author: Abhijit Dasgupta, PhD
date: Spring 2019

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = '#> ', message = F, warning = F, cache=F)
library(tidyverse)
theme_set(theme_bw())
```

layout: true

<div class="my-header">
<span>BIOF 439, Spring 2019</span></div>


---
class:middle, center, inverse

# Annotations

---

## Stand-alone stories

- You would like a data visualization to stand on its own
- Relevant information should be placed on the graph
- However, you need to balance the information content with real estate
    - Don't clutter the graph and make it not readable

---
    

## An example 

<div id="origecon"></div>
![](img/economist1.gif)

---

## Reconstructing this annotated graph

.pull-left[
```{r a1, eval = F, echo = T, message=F}
library(tidyverse)
econ_data <- rio::import('data/EconomistData.csv')
ggplot(econ_data,
       aes(x = CPI, y = HDI, color=Region))+
  geom_point()
```
]
.pull-right[
```{r 05-Lecture5-1, eval = T, echo = F, ref.label = "a1"}
```
]

---

## Reconstructing this annotated graph

.pull-left[
```{r a2, eval = F, echo = T}
ggplot(econ_data,
       aes(x = CPI, y = HDI, color=Region))+
  geom_point() +
  geom_smooth(color='red', se=F)
```

#### Add a trend line
]
.pull-right[
```{r 05-Lecture5-2, eval = T, echo = F, ref.label = "a2", message=F}
```
]

---

## Reconstructing this annotated graph

.pull-left[
```{r a3, eval = F, echo = T, message=F}
ggplot(econ_data,
       aes(x = CPI, y = HDI, color=Region))+
  geom_smooth(color='red', se=F) + #<<
  geom_point()#<<
```

#### Reverse order so points are above line
]
.pull-right[
```{r 05-Lecture5-3, eval = T, echo = F, ref.label = "a3"}
```
]

---

## Reconstructing this annotated graph

.pull-left[
```{r a4, eval = F, echo = T}
ggplot(econ_data,
       aes(x = CPI, y = HDI, color=Region))+
  geom_smooth(color='red', se=F) + 
  geom_point(shape = 1, size = 4, stroke=1.25)#<<

```

#### Different shape for points
]
.pull-right[
```{r 05-Lecture5-4, eval = T, echo = F, ref.label = "a4"}
```
]

---

## Reconstructing this annotated graph

.pull-left[
```{r a5, eval = F, echo = T}
pointsToLabel <- c("Russia", "Venezuela", "Iraq", "Myanmar", "Sudan",
                   "Afghanistan", "Congo", "Greece", "Argentina", "Brazil",
                   "India", "Italy", "China", "South Africa", "Spane",
                   "Botswana", "Cape Verde", "Bhutan", "Rwanda", "France",
                   "United States", "Germany", "Britain", "Barbados", "Norway", "Japan",
                   "New Zealand", "Singapore")
ggplot(econ_data,
       aes(x = CPI, y = HDI, color=Region))+
  geom_smooth(color='red', se=F) + 
  geom_point(shape = 1, size = 4, stroke=1.25) + 
  geom_text(aes(label=Country),
            color = 'gray20')
```

#### Label countries
]
.pull-right[
```{r 05-Lecture5-5, eval = T, echo = F, ref.label = "a5"}
```
]

---

## Reconstructing this annotated graph

.pull-left[
```{r a6, eval = F, echo = T}
pointsToLabel <- c("Russia", "Venezuela", "Iraq", "Myanmar", "Sudan",
                   "Afghanistan", "Congo", "Greece", "Argentina", "Brazil",
                   "India", "Italy", "China", "South Africa", "Spane",
                   "Botswana", "Cape Verde", "Bhutan", "Rwanda", "France",
                   "United States", "Germany", "Britain", "Barbados", "Norway", "Japan",
                   "New Zealand", "Singapore")
ggplot(econ_data,
       aes(x = CPI, y = HDI, color=Region))+
  geom_smooth(color='red', se=F) + 
  geom_point(shape = 1, size = 4, stroke=1.25) + 
  geom_text(aes(label=Country),
            color = 'gray20', 
            data = econ_data %>% #<<
              filter(Country %in% pointsToLabel))#<<
```

#### Better, but labels are overlayed on points
]
.pull-right[
```{r 05-Lecture5-6, eval = T, echo = F, ref.label = "a6"}
```
]

---

## Reconstructing this annotated graph

.pull-left[
```{r a7, eval = F, echo = T}
library(ggrepel) #<<
pointsToLabel <- c("Russia", "Venezuela", "Iraq", "Myanmar", "Sudan",
                   "Afghanistan", "Congo", "Greece", "Argentina", "Brazil",
                   "India", "Italy", "China", "South Africa", "Spane",
                   "Botswana", "Cape Verde", "Bhutan", "Rwanda", "France",
                   "United States", "Germany", "Britain", "Barbados", "Norway", "Japan",
                   "New Zealand", "Singapore")
(plt <- ggplot(econ_data,
       aes(x = CPI, y = HDI, color=Region))+
  geom_smooth(color='red', se=F) + 
  geom_point(shape = 1, size = 4, stroke=1.25) + 
  geom_text_repel(aes(label=Country),#<<
            color = 'gray20', 
            force=20,
            data = econ_data %>% 
              filter(Country %in% pointsToLabel)))
```

]
.pull-right[
```{r 05-Lecture5-7, eval = T, echo = F, ref.label = "a7"}
```
]

---

## Reconstructing this annotated graph

.pull-left[

Let's re-order the regions

```{r a8, eval = F, echo = T}
econ_data$Region <- 
  factor(econ_data$Region,
         levels = c("EU W. Europe",
                    "Americas",
                    "Asia Pacific",
                    "East EU Cemt Asia",
                    "MENA",
                    "SSA"),
         labels = c("OECD",
                    "Americas",
                    "Asia &\nOceania",
                    "Central &\nEastern Europe",
                    "Middle East &\nnorth Africa",
                    "Sub-Saharan\nAfrica"))

plt$data = econ_data #<<
plt #<<
```
]
.pull-right[
```{r 05-Lecture5-8, eval = T, echo = F, ref.label = "a8"}
```
]

---

## Reconstructing this annotated graph

.pull-left[
Clean up the graphic
```{r a9, eval = F, echo = T}
(plt_corrupt <- 
   plt +
  scale_x_continuous(name = 'Corruption Perceptions Index',
                     breaks = 1:10) +
  scale_y_continuous(name="Human Development Index",
                     breaks = seq(0.2, 1, by = 0.1))+
  scale_color_manual(name = '',
                     values = c("#24576D",
                                "#099DD7",
                                "#28AADC",
                                "#248E84",
                                "#F2583F",
                                "#96503F")) +
  ggtitle("Corruption and Human development")+
  theme_bw()+
  theme(legend.position='top',
        legend.direction='horizontal')
)
```
]
.pull-right[
```{r 05-Lecture5-9, eval = T, echo = F, ref.label = "a9"}
```
]


---
class:middle,center

# Adding derived statistics to a plot

---

## Adding group means

.pull-left[
```{r b1, eval = F, echo = T}
ggplot(iris, 
       aes(x = Sepal.Length,
           y = Sepal.Width,
           color = Species))+
  geom_point()+
  theme_bw()
```
]
.pull-right[
```{r 05-Lecture5-10, eval = T, echo = F, ref.label = "b1"}
```
]

---

## Adding group means

.pull-left[
```{r 05-Lecture5-11, eval = T, echo = T}
means <- iris %>% group_by(Species) %>% 
  summarize_at(vars(starts_with('Sepal')),
               mean)
means
```


```{r b2, eval = F, echo = T}
ggplot(iris, 
       aes(x = Sepal.Length,
           y = Sepal.Width,
           color = Species))+
  geom_point() + 
  geom_point(data = means, #<<
             size=5) +#<<
  theme_bw()
```


```{r 05-Lecture5-12, eval = T, echo = T}
```
]
.pull-right[
```{r 05-Lecture5-13, eval = T, echo = F, ref.label = "b2"}
```
]
---

## Adding regression metrics

.pull-left[

Regress highway mileage on city mileage (data: mpg)

```{r c1, eval = F, echo = T}
mod1 <- lm(hwy ~ cty, data = mpg)
r2 <- broom::glance(mod1) %>% pull(r.squared)

ggplot(mpg, 
       aes(x = cty, y = hwy))+
  geom_point() + 
  geom_smooth(method = 'lm', se=F) +
  theme_bw()
```
]
.pull-right[
```{r 05-Lecture5-14, eval = T, echo = F, ref.label = "c1"}
```
]

---

## Adding regression metrics

.pull-left[

Regress highway mileage on city mileage (data: mpg)

```{r c2, eval = F, echo = T}
mod1 <- lm(hwy ~ cty, data = mpg)
r2 <- broom::glance(mod1) %>% pull(r.squared) %>% 
  round(., 2)

ggplot(mpg, 
       aes(x = cty, y = hwy))+
  geom_point() + 
  geom_smooth(method = 'lm', se=F)+
  annotate(geom='text',
           x = 15, y = 40,
           label=glue::glue("R^2 == {r}",r=r2),
           size=12,
           parse=T) + 
  theme_bw()
```
]
.pull-right[
```{r 05-Lecture5-15, eval = T, echo = F, ref.label = "c2"}
```
]

---

## Highlighting regions

.pull-left[
```{r d1, eval = F, echo = T}
mpg %>% 
  mutate(cyl = as.factor(cyl)) %>% 
  ggplot(aes(x = cyl, y = hwy)) + 
  geom_boxplot() + 
  theme_bw()
```
]
.pull-right[
```{r 05-Lecture5-16, eval = T, echo = F, ref.label = "d1"}
```
]

---

## Highlighting regions

.pull-left[
```{r d2, eval = F, echo = T}
mpg %>% 
  mutate(cyl = as.factor(cyl)) %>% 
  ggplot(aes(x = cyl, y = hwy)) + 
  geom_boxplot() + 
  theme_bw()+
  annotate(geom = 'rect',
           xmin=3.75,xmax=4.25,
           ymin = 22, ymax = 28,
           fill = 'red',
           alpha = 0.2) +
  annotate('text', 
           x = 4.5, y = 25,
           label = 'Outliers?',
           hjust = 0)+
  coord_cartesian(xlim = c(0,5))+ 
  theme_bw()
```
]
.pull-right[
```{r 05-Lecture5-17, eval = T, echo = F, ref.label = "d2"}
```
]


---
class:center, middle,inverse

# Maps

---

For maps, we need a couple of new packages. 

- `sf`: Simple features in R
- `rnaturalearth` & `rnaturalearthdata`: map data

---

.pull-left[
```{r e1, eval = F, echo = T}
library(sf)
library(rnaturalearth)
library(rnaturalearthdata)

world <- ne_countries(scale='medium', returnclass='sf')
ggplot(data = world) + 
  geom_sf()
```
]
.pull-right[
```{r 05-Lecture5-18, eval = T, echo = F, ref.label = "e1"}
```
]

---

.pull-left[
```{r e2, eval = F, echo = T}
library(sf)
library(rnaturalearth)
library(rnaturalearthdata)

world <- ne_countries(scale='medium', returnclass='sf')
ggplot(data = world) + 
  geom_sf(aes(fill = pop_est))
```
]
.pull-right[
```{r 05-Lecture5-19, eval = T, echo = F, ref.label = "e2"}
```
]

---

.pull-left[
```{r e3, eval = F, echo = T}
library(sf)
library(rnaturalearth)
library(rnaturalearthdata)

world <- ne_countries(scale='medium', returnclass='sf')
ggplot(data = world) + 
  geom_sf(aes(fill = pop_est))+
  scale_fill_viridis_c(option = 'plasma', trans='sqrt')
  
```
]
.pull-right[
```{r 05-Lecture5-20, eval = T, echo = F, ref.label = "e3"}
```
]

---

## Looking at Florida

.pull-left[
```{r f1, eval = F, echo = T}
library(maps)
states <- st_as_sf(maps::map('state', plot = F, fill = T)) %>% 
  cbind(st_coordinates(st_centroid(.))) %>% 
  mutate(ID = str_to_title(ID))

ggplot(data = world)+
  geom_sf() + 
  geom_sf(data = states, fill = NA) + 
  geom_text(data = states, aes(X, Y, label = ID), 
            size = 5) + 
  coord_sf(xlim = c(-88, -78), ylim = c(24.5, 33), 
           expand = F)
```
]
.pull-right[
```{r 05-Lecture5-21, eval = T, echo = F, ref.label = "f1", warning=F}
```
]

---

## Looking at Florida

.pull-left[
```{r f2, eval = F, echo = T}
library(maps)
states <- st_as_sf(map('state', plot = F, fill = T)) %>% 
  cbind(st_coordinates(st_centroid(states))) %>% 
  mutate(ID = str_to_title(ID))

ggplot(data = world)+
  geom_sf() + 
  geom_sf(data = states, fill = NA) + 
  geom_label(data = states, aes(X, Y, label = ID), #<<
            size = 5) + 
  coord_sf(xlim = c(-88, -78), ylim = c(24.5, 33), 
           expand = F)
```
]
.pull-right[
```{r 05-Lecture5-22, eval = T, echo = F, ref.label = "f2", warning=F}
```
]

---

## Looking at the Florida elections

.pull-left[
```{r g1, eval = F, echo = T}
source('data/florida.R')
head(florida_election)
florida_election <- florida_election %>% 
  mutate_at(vars(ends_with('perc')), 
            ~as.numeric(str_remove(., '%')))

```
]
.pull-right[
```{r 05-Lecture5-23, eval = T, echo = F, ref.label = "g1"}
```
]

---

## Looking at the Florida election

.pull-left[
Now we need the map information
```{r g2, eval = F, echo = T}
library(maps)
counties <- st_as_sf(maps::map('county', plot = F, fill = T))
head(counties)
counties <- counties %>% filter(str_detect(ID, 'florida'))
counties <- counties %>% separate(ID, c('State','County'), sep = ',') %>% 
  mutate_at(vars(State:County), str_to_title)
```
]
.pull-right[
```{r 05-Lecture5-24, eval = T, echo = F, ref.label = "g2"}
```
]

---

## Looking at the Florida election

.pull-left[
The nice thing about the `sf` package is that it renders all the data into a data frame, so adding to it, or merging new data becomes easy. 

We will now merge the election data with the map data

```{r g3, eval = F, echo = T}
election_by_county <- counties %>% left_join(florida_election)
head(election_by_county)

```
]
.pull-right[

```{r 05-Lecture5-25, eval = T, echo = F, ref.label = "g3"}
```
]

---

## Looking at the Florida election

.pull-left[
Now we're ready to plot

```{r g4, eval = F, echo = T}
ggplot(election_by_county) + 
  geom_sf(aes(fill = Gore_perc)) + 
  scale_fill_viridis_c(option = 'plasma') + 
  labs(caption = 'Source: Wikipedia')
```
]
.pull-right[
```{r 05-Lecture5-26, eval = T, echo = F, ref.label = "g4"}
```
]

---

## Looking at the Florida election

.pull-left[

We can clean this up a bit, and add surrounding states

```{r g5, eval = F, echo = T}
(plt_map <- ggplot(data = world)+
  geom_sf() + 
  geom_sf(data = states, fill = NA) + 
  geom_sf(data = election_by_county, aes(fill=Gore_perc)) + 
  geom_label(data = states %>% filter(ID != 'Florida'), aes(X, Y, label = ID), #<<
            size = 5) +
  coord_sf(xlim = c(-88, -78), ylim = c(24.5, 33), 
           expand = F) +
  labs(x = '', y = '', fill = 'Percentage for Gore') + 
  scale_fill_viridis_c(option = 'plasma'))
```
]
.pull-right[
```{r 05-Lecture5-27, eval = T, echo = F, ref.label = "g5"}
```
]

---
class: middle,center

# Genomic data

---

## Visualizing a proteomic network

.pull-left[
We read a dataset that contains the network relationships between different proteins
```{r h1, eval = F, echo = T}
library(ggnetwork)
datf <- rio::import('data/string_graph.txt')
head(datf)
```
]
.pull-right[
```{r 05-Lecture5-28, eval = T, echo = F, ref.label = "h1"}
```
]

---

## Visualizing a proteomic network

.pull-left[
The `igraph` package allows the creation of network graphs. 

However, here, we're only using it for data ingestion
```{r h2, eval = F, echo = T}
grs <- igraph::graph_from_data_frame(datf[,c('node1','node2')], 
                                     directed = F)
grs
```

We see that this object holds the different connections. 
]
.pull-right[
```{r 05-Lecture5-29, eval = T, echo = F, ref.label = "h2"}
```
]

---

## Visualizing a proteomic network

.pull-left[
We can then transform this data into `ggplot`-friendly data, to use `ggplot` for the plotting


```{r h3, eval = F, echo = T}
ggdf <- ggnetwork(grs, layout = 'fruchtermanreingold')

ggplot(ggdf, aes(x = x, y = y, xend = xend, yend = yend)) +
  geom_edges(color = "black", #<<
             curvature = 0.1, size = 0.95, alpha = 0.8)+ #<<
  geom_nodes(aes(x = x, y = y), #<<
             size = 3, #<<
             alpha = 0.5, #<<
             color = "orange") + #<<
  geom_nodelabel_repel(aes(label = vertex.names), #<<
                     size=4, color="#8856a7") +#<<
  theme_blank() + theme(legend.position = "none")
```
]
.pull-right[
```{r 05-Lecture5-30, eval = T, echo = F, ref.label = "h3"}
```
]

---
class: middle, center

# Composing different genomic data into tracks

---

## The `ggbio` package

The `ggbio` package has several functions that allow graphical representations of different genomic entities.

.pull-left[

An ideogram

```{r j1, eval = F, echo = T}
library(ggbio)
p.ideo <- Ideogram(genome = 'hg19')
p.ideo
```
]
.pull-right[
```{r 05-Lecture5-31, eval = T, echo = F, ref.label = "j1"}
```
]

---

## The `ggbio` package

.pull-left[

Visualizing the gene model

```{r j2, eval = F, echo = T}
library(Homo.sapiens)
data(genesymbol, package='biovizBase')
wh <- genesymbol[c('BRCA1','NBR1')]
wh <- range(wh, ignore.strand=T)
p.txdb <- autoplot(Homo.sapiens, which = wh)
p.txdb
```
]
.pull-right[
```{r 05-Lecture5-32, eval = T, echo = F, ref.label = "j2"}
```
]

---

## The `ggbio` package

.pull-left[

Putting it into tracks

```{r j3, eval = F, echo = T}
library(GenomicRanges)
gr17 <- GRanges("chr17", IRanges(41234415, 41234569))
tks <-  tracks(p.ideo, gene = p.txdb) + xlim(gr17)
tks + theme_tracks_sunset()
```
]
.pull-right[
```{r 05-Lecture5-33, eval = T, echo = F, ref.label = "j3"}
```
]

---
class: middle, center

# Interactive maps

---

## Using plotly

.pull-left[
```{r k1, eval = F, echo = T}
library(plotly)

ggplotly(plt_map)
```
]
.pull-right[
```{r 05-Lecture5-34, eval = T, echo = F, ref.label = "k1"}
```
]

---

## Using real maps

.pull-left[
```{r k2, eval = F, echo = T}
library(leaflet)

m <- leaflet() %>%
  addTiles() %>%  # Add default OpenStreetMap map tiles
  addMarkers(lng=174.768, lat=-36.852, popup="The birthplace of R")
m 
```
]
.pull-right[
```{r 05-Lecture5-35, eval = T, echo = F, ref.label = "k2"}
```
]

---
class: center, middle

# Heatmaps


---

## Recall our heatmap

.pull-left[
```{r l1, eval = F, echo = T}
library(Biobase)
#data(sample.ExpressionSet)
exdat <-  readRDS('data/exprset.rds')
library(limma)
design1 <- model.matrix(~type, data=pData(exdat))
lm1 <- lmFit(exprs(exdat), design1)
lm1 <- eBayes(lm1) # compute linear model for each probeset
geneID <- rownames(topTable(lm1, coef = 2, number = 100, 
                            adjust.method = 'none',
                            p.value = 0.05))
exdat2 <- exdat[geneID,] # Keep features with p-values < 0.05

library(Heatplus)
reg1 <- regHeatmap(exprs(exdat2), legend=2, col=heat.colors, 
                   breaks=-3:3)
plot(reg1)
```
]
.pull-right[
```{r 05-Lecture5-36, eval = T, echo = F, ref.label = "l1"}
```
]

---

## Using `heatmaply`

.pull-left[
```{r l2, eval = F, echo = T}
library(heatmaply)
heatmaply(exprs(exdat2))

```
]
.pull-right[
```{r 05-Lecture5-37, eval = T, echo = F, ref.label = "l2"}
```
]

---

## Using heatmaply

.pull-left[
You can also easily use the correlation metric

```{r l3, eval = F, echo = T}
heatmaply(exprs(exdat2), distfun = 'pearson')
```
]
.pull-right[
```{r 05-Lecture5-38, eval = T, echo = F, ref.label = "l3"}
```
]

---

## Using `d3heatmap`

.pull-left[
```{r l4, eval = F, echo = T}
library(d3heatmap)
d3heatmap(exprs(exdat2))

```
]
.pull-right[
```{r 05-Lecture5-39, eval = T, echo = F, ref.label = "l4"}
```
]

---
class: middle, center

# Manhattan plots

---

```{r 05-Lecture5-40}
library(qqman) #<<
manhattan(gwasResults)
```

---

```{r 05-Lecture5-41}
library(manhattanly)
manhattanly(gwasResults)
```

---

## Interactions using plotly

- Plot.ly is a web service that produces interactive graphics from data
- They made their backend open-source
- In R, you can interact with plot.ly using the package `plotly`.

---

## Interactions using plotly

- Plotly makes it very easy to create interactive plots based on ggplot

.pull-left[
```{r m1, eval = F, echo = T}
plt_corrupt
```
]
.pull-right[
```{r 05-Lecture5-42, eval = T, echo = F, ref.label = "m1"}
```
]

---

## Interactions using plotly

- Plotly makes it very easy to create interactive plots based on ggplot

.pull-left[
```{r m2, eval = F, echo = T}
library(plotly)

ggplotly(plt_corrupt)
```
]
.pull-right[
```{r 05-Lecture5-43, eval = T, echo = F, ref.label = "m2"}
```
]
